pkg/seedadmission/pods.go

```diff
@@ -66,67 +66,110 @@ func MutatePod(ctx context.Context, c client.Client, logger *logrus.Logger, requ
 
 // mutateShootControlPlanePodAnnotations checks if the object mutation is allowed and return the needed patches .
 func mutateShootControlPlanePodAnnotations(ctx context.Context, c client.Client, logger *logrus.Entry, pod runtime.Object, namespace string) ([]jsonpatch.JsonPatchOperation, error) {
-	var patch []jsonpatch.JsonPatchOperation
-	// determine whether to perform mutation
-	isMutationRequired, err := mutationRequired(ctx, c, namespace)
+	desiredAnnotations := map[string]string{"fluentbit.io/exclude": "true"}
+
+	acc, err := meta.Accessor(pod)
 	if err != nil {
 		return nil, err
 	}
 
-	acc, err := meta.Accessor(pod)
+	shoot, err := extractShoot(ctx, c, namespace)
 	if err != nil {
 		return nil, err
 	}
-
-	if !isMutationRequired {
-		logger.Debugf("Skipping mutation for %s/%s due to policy check", acc.GetNamespace(), acc.GetName())
+	if shoot == nil {
+		logger.Debugf("Skipping mutation for %s/%s because shoot could not be extracted", acc.GetNamespace(), acc.GetName())
 		return nil, nil
 	}
 
-	originalAnnotations := acc.GetAnnotations()
-	annotationsToAdd := map[string]string{
-		"fluentbit.io/exclude": "true",
+	if mustAddOrReplaceAnnotations(shoot) {
+		return computeAddOrReplacePatch(acc.GetAnnotations(), desiredAnnotations), nil
 	}
 
-	if originalAnnotations != nil {
-		for key, value := range annotationsToAdd {
-			patch = append(patch, jsonpatch.JsonPatchOperation{
-				Operation: "replace",
-				Path:      "/metadata/annotations/" + strings.ReplaceAll(key, "/", "~1"),
-				Value:     value,
-			})
-		}
-		return patch, nil
+	if mustRemoveAnnotations(acc.GetAnnotations(), desiredAnnotations) {
+		return computeRemovePatch(acc.GetAnnotations(), desiredAnnotations), nil
 	}
 
-	patch = append(patch, jsonpatch.JsonPatchOperation{
-		Operation: "add",
-		Path:      "/metadata/annotations",
-		Value:     annotationsToAdd,
-	})
-	return patch, nil
+	logger.Debugf("Skipping mutation for %s/%s due to policy check", acc.GetNamespace(), acc.GetName())
+	return nil, nil
 }
 
-func mutationRequired(ctx context.Context, c client.Client, namespace string) (bool, error) {
+func extractShoot(ctx context.Context, c client.Client, namespace string) (*gardencorev1beta1.Shoot, error) {
 	cluster, err := extensionscontroller.GetCluster(ctx, c, namespace)
```

Comment:
I know, this was already here. But didn't you say, we want to avoid references from g/pkg to g/extensions?
Can you fix that here?

Reply:
What do you suggest? If we move the `cluster.go` file from `extensionscontroller` package to somewhere else we have to adapt all extension controllers. I don't want to copy the logic of getting the shoot out of a cluster.

Reply:
What about moving the logic to somewhere in `g/pkg` and make `extensionscontroller.GetCluster` an alias to that function?
Then we don't need to copy the logic, don't need to adapt the extensions but can still use it in `g/pkg` without importing `g/extensions`.

This would work, but I would also be fine with it, if we agree on having these imports (and probably also adding more in the future). I guess, we anyways don't want to split the whole thing up again 😄 
Just wanted to double-check and have no strong opinion on this topic to be honest.

Reply:
Hm yes, we don't plan to split it again. The problem is that g/gext was developed side-by-side to g/g and is now just "copied" in g/g again. Over time, I would like to see the `g/extensions` folder disappear/to be merged into `g/pkg`.

Many things in `g/extensions` can be merged with `g/pkg` in some way. A good example is the `g/extensions/gardener/chartrenderer`. It's also strange that we have `g/extensions/log` and `g/pkg/logger`. Eventually, this should be harmonized.

To get started, I've moved `g/extensions/controller/cluster.go` to `g/pkg/extensions/cluster.go` and made type/func aliases (just to not make this PR too large; the `g/extensions/controller/cluster.go` should be deleted in a subsequent PR and the usages should be adapted to `g/pkg/extensions`.

PTAL! :)

Reply:
Yes. There are some duplications and other stuff we have to harmonize.
Thanks for adapting and starting with this change! :)